config-reloader/datasource/kubedatasource/fluentdconfig.go

```diff
@@ -0,0 +1,223 @@
+
+	fdDS := &FluentdConfigDS{
+		cfg:        cfg,
+		fdlist:     fluentdConfigLister,
+		fdready:    factory.Logs().V1beta1().FluentdConfigs().Informer().HasSynced,
+		updateChan: updateChan,
+	}
+
+	factory.Logs().V1beta1().FluentdConfigs().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: fdDS.handleFDChange,
+		UpdateFunc: func(old, new interface{}) {
+			fdDS.handleFDChange(new)
+		},
+		DeleteFunc: fdDS.handleFDChange,
+	})
+
+	// Verify CRD availability
+	if err := fdDS.checkAndInstallCRDs(kubeCfg); err != nil {
+		return nil, err
+	}
+
+	factory.Start(nil)
+
+	return fdDS, nil
+}
+
+// IsReady returns a boolean specifying whether the FluentdConfigDS is ready
+func (f *FluentdConfigDS) IsReady() bool {
+	return f.fdready()
+}
+
+// GetFluentdConfig returns the fluentd configs for the given ns extracted
+// by the configured FluentdConfigs k8s resources
+func (f *FluentdConfigDS) GetFluentdConfig(namespace string) (string, error) {
+	// Grab all FluentdConfigs k8s resources in the given ns
+	fluentdConfigs, err := f.fdlist.FluentdConfigs(namespace).List(labels.Everything())
+	if err != nil {
+		return "", err
+	}
+
+	// Extract fluentd
+	configData := make([]string, 0, len(fluentdConfigs))
+	for _, fd := range fluentdConfigs {
+		configData = append(configData, fd.Spec.FluentConf)
+	}
+
+	// Concatenate all namespace's configs
+	return strings.Join(configData, "\n"), nil
+}
+
+// handleFDChange reacts to changes in the FluentdConfigs k8s resources and notifies the
+// main controller to re-run the main loop and sync the state
+func (f *FluentdConfigDS) handleFDChange(obj interface{}) {
+	// If the controller is monitoring all namespaces, it needs to react
+	// to all notifications of changes to FluentdConfigs resources.
+	// If instead only a subset of namespaces is being monitored, there
+	// is no need to run the control loop unless the changed FluentdConfig
+	// resource is in one of the monitored namespaces
+	if len(f.cfg.Namespaces) != 0 {
+		var object metav1.Object
+		var ok bool
+		if object, ok = obj.(metav1.Object); !ok {
+			tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+			if !ok {
+				logrus.Warnf("error decoding object, invalid type")
+				return
+			}
+			object, ok = tombstone.Obj.(metav1.Object)
+			if !ok {
+				logrus.Warnf("error decoding object tombstone, invalid type")
+				return
+			}
+		}
+
+		toProcess := false
+		for _, ns := range f.cfg.Namespaces {
+			if object.GetNamespace() == ns {
+				toProcess = true
+				break
+			}
+		}
+		if !toProcess {
+			return
+		}
+	}
+
+	select {
+	case f.updateChan <- time.Now():
+	default:
+		// There is already one pending notification. Useless to send another one since, when
+		// the pending one will be processed all new changes will be reloaded.
+	}
+}
+
+///////////////////////////////////////////
+// CRD Management for autoinstall
+///////////////////////////////////////////
+
+var fluentdConfigCRD = v1.CustomResourceDefinition{
```

Comment:
@tommasopozzetti  This is a great PR. I think v1 for crd is supported from 1.16 onwards. Can we handle this based on a feature flag? or do you think any better approach to handle between v1 and v1beta1?
I observed this when I pulled this change and tested it locally. In worst-case scenario, if those crds are created outside the code, can that be provided as a workaround?

Since this a major feature, can you also consider a doc update ?

Reply:
Agreed, we need to ensure it doesn't fail on older Kubernetes versions prior to 1.16. 

> The config-reloader has been equipped with the capability of installing the CRD at startup if requested, so no manual actions to enable it on the cluster are needed.

We should definitely have the ability to turn this off and do it manually. It's an interesting approach to install the CRDs from the code but we need to decide if this is something we really want to do longer term - if we install from the code we also need to maintain/upgrade in code vs a YAML file. This has the advantage of having less loosely coupled dependencies, I guess it depends how frequently we see that CRD changing.

Reply:
Thank you for the feedback.
This approach to CRDs is similar to the one used by coreos/dex. I like it as it simplifies the process of managing CRDs for the user, it continuously monitors CRD availability at each restart of the operator and it couples the CRD definition to the code, allowing them to easily evolve together.
In case of CRD evolution, the code can evolve to be able to apply updates if required automatically, again taking a possible source of errors out of the hands of the users.
For the problem of Kubernetes versions preceding 1.16, I think we can add support also for v1beta1 and automatically choose the correct version based on the environment.
We could also decide to provide this auto-management of CRDs as a controllable feature that we can activate through the chart. This will however require to maintain also the CRD yaml file so that users can apply it manually (or through the helm chart) if this auto-management feature is not active.
Let me know your thoughts and I'll be happy to add the needed modifications to this PR.

Reply:
I am fine with the approach to handle both the versions in the code. Since it performs an existence check, this wont be stopping the users from maintaining it outside. I believe your other commit #110 will clear the old configs, if we toggle the behaviour.

@carpenterm Any thoughts ?

Reply:
@dimalbaby @carpenterm I pushed the commit adding the capability of automatically choosing the right version of the CRD API based on the system.
Unfortunately I don't have a Kubernetes cluster with version < 1.16 handy at the moment, so could you could help with testing it in that environment?

Reply:
@tommasopozzetti Thanks for your effort for making this change. I can test it on kubernetes cluster with version < 1.16. I will update you by early next week.

Reply:
@dimalbaby I pushed one more commit to refactor the new crd management code to reduce code duplication. It should not affect the behavior.

Reply:
@tommasopozzetti I cherry picked the changes and hit a minor issue while running it using  vibeta1. I will submit a commit to the pull request. Please take a look

Reply:
@tommasopozzetti I have added a commit https://github.com/vmware/kube-fluentd-operator/pull/111/commits/38e6ad256c23a37d3536d775013c1dfe15e98a8d. Overall the changes looks good to me. I have performed a basic testing with the above changes and it worked really well. Thanks for addressing the v1beta1 support
The only question is how customers will gradually migrate to crds if they are using config map approach today. I think it would be nice to have a migration plan where the reloader can pick both the configs and later switch to only crds. Any thoughts ? Probably we can consider it separately

Reply:
I tend to agree, we run clusters with lots of different service teams working on different products sharing the same kube-fluentd-operator instance and I'm assuming that others are also in that situation. Migrating all of the teams to CRDs at the same time we switch the server to CRDs would be a bit of a headache, it would be better if we could have both CRDs and ConfigMaps enabled for a while to enable teams to migrate. Would be great to dynamically figure out which one to use based on whether a ConfigMap or CRD exists, or if that's too much work, perhaps they could control it themselves via an annotation on the namespace and the cluster administrator could set a default?

Reply:
Thank you for testing and for fixing the bug @dimalbaby 

We could look at creating an extra "virtual" datasource that basically internally manages an instance of the configmap datasource and an instance of the crd datasource and concats the fluentd configs from both. This would allow running both datasources side by side. This virtual datasource can then be deprecated and possibly removed in a later release if there is interest in dropping support for configmaps or for migration (this is mostly to avoid the complexity of managing an extra datasource as the operator evolves). Let me know your thoughts.

Reply:
I agree @tommasopozzetti. But what would happen if there is collision- loading the configs from both the sources? Probably the CRDs will override ?

Reply:
I think that it should be left up to who is managing the namespace to handle the issue. I see it as the same issue of what would happen if in the `multimap` datasource multiple configmaps are defined with collisions.
The datasource simply provides the input, and this new virtual datasource would allow providing inputs from two sources at the same time, but it is left up to who is writing that input to provide a correct one. If a collision generates a syntactic error, fluentd validation will catch it. If, instead, it creates a semantic error, then it is allowed, just as it would be allowed in the `multimap` datasource or even in the default datasource in case that same concatenated config is put in one configmap.
The correct migration process would require first installing the new release and enabling this `migration mode`. Then, one by one, the configmaps should be replaced by the corresponding CRDs. This coexistence will allow this process to last for as long as needed to get all teams up to speed, and finally the datasource can be switched to pure CRD, thus disabling configmaps alltogether.

Reply:
That make sense. Thanks @tommasopozzetti